db/inlineskiplist.h

```diff
@@ -549,43 +665,103 @@ void InlineSkipList<Comparator>::Insert(const char* key) {
   }
 
   // Our data structure does not allow duplicate insertion
   assert(prev_[0]->Next(0) == nullptr || !Equal(key, prev_[0]->Next(0)->Key()));
 
-  // Find the Node that we placed before the key in AllocateKey
-  Node* x = reinterpret_cast<Node*>(const_cast<char*>(key)) - 1;
-  int height = x->UnstashHeight();
-  assert(height >= 1 && height <= kMaxHeight_);
+  for (int i = 0; i < height; i++) {
+    x->InsertAfter(prev_[i], i);
+  }
+  prev_[0] = x;
+  prev_height_.store(height, std::memory_order_relaxed);
+}
 
-  if (height > GetMaxHeight()) {
-    for (int i = GetMaxHeight(); i < height; i++) {
-      prev_[i] = head_;
+template <class Comparator>
+void InlineSkipList<Comparator>::FindWithHint(
+    const char* key, InsertHint* hint, int start_level, int stop_level) const {
+  assert(start_level >= stop_level);
+  int level = start_level;
+  while (level < GetMaxHeight() &&
+      !KeyIsNextToNode(key, hint->prev[level], level)) {
+    level++;
+  }
+  if (level == start_level) {
+    while (level > stop_level &&
+        KeyIsNextToNode(key, hint->prev[level - 1], level - 1)) {
+      level--;
     }
-
-    // It is ok to mutate max_height_ without any synchronization
-    // with concurrent readers.  A concurrent reader that observes
-    // the new value of max_height_ will see either the old value of
-    // new level pointers from head_ (nullptr), or a new value set in
-    // the loop below.  In the former case the reader will
-    // immediately drop to the next level since nullptr sorts after all
-    // keys.  In the latter case the reader will use the new node.
-    max_height_.store(height, std::memory_order_relaxed);
   }
+  if (level > stop_level) {
+    FindLessThan(key, hint->prev, hint->prev[level], level, stop_level);
+  }
+}
 
-  for (int i = 0; i < height; i++) {
-    // NoBarrier_SetNext() suffices since we will add a barrier when
-    // we publish a pointer to "x" in prev[i].
-    x->NoBarrier_SetNext(i, prev_[i]->NoBarrier_Next(i));
-    prev_[i]->SetNext(i, x);
+template <class Comparator>
+void InlineSkipList<Comparator>::InsertWithHint(
+    const char* key, void** hint_ptr) {
+  int height = 0;
+  Node* x = GetNodeForInsert(key, &height);
+  MaybeInvalidatePrev(height);
+  
+  assert(hint_ptr != nullptr);
+  InsertHint* hint = reinterpret_cast<InsertHint*>(*hint_ptr);
+  if (hint == nullptr) {
+    hint = AllocateInsertHint();
+    *hint_ptr = reinterpret_cast<void*>(hint);
+  }
+  int level = 0;
+  for (; level < hint->num_levels; level++) {
+    if (KeyIsAfterNode(key, hint->prev[level])) {
+      break;
+    }
+  }
+  if (level >= hint->num_levels) {
+    FindWithHint(key, hint, std::max<int>(level, height - 1), level);
+  } else {
+    int hint_max_height = 0;
+    for (int i = level; i < hint->num_levels; i++) {
+      hint_max_height = std::max<int>(hint_max_height, hint->prev_height[i]);
+    }
+    if (height > hint_max_height) {
+      FindWithHint(key, hint, height - 1, hint_max_height);
```

Comment:
We are looking for prev from root, right? Can we write more specific code for that, rather than a general `FindWithHint()`?